common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/PKCEChallenge.java

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.providers.oauth2;
+
+import android.util.Base64;
+
+import com.microsoft.identity.common.exception.ClientException;
+import com.microsoft.identity.common.exception.ErrorStrings;
+
+import java.io.UnsupportedEncodingException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+/**
+ * The client creates a code challenge derived from the code
+ * verifier by using one of the following transformations.
+ * <p>
+ * Sophisticated attack scenarios allow the attacker to
+ * observe requests (in addition to responses) to the
+ * authorization endpoint.  The attacker is, however, not able to
+ * act as a man in the middle. To mitigate this,
+ * "code_challenge_method" value must be set either to "S256" or
+ * a value defined by a cryptographically secure
+ * "code_challenge_method" extension. In this implementation "S256" is used.
+ * <p>
+ * Example for the S256 code_challenge_method
+ *
+ * @see <a href="https://tools.ietf.org/html/rfc7636#page-17">RFC-7636</a>
+ */
+
+public class PKCEChallenge {
```

Comment:
I think I may have written this back in the day. Blame me for the capitalization, but let's change it